File: 백준/Gold/32203. 연락/연락.py
#Authored by : chjh2129
import sys

input = sys.stdin.readline

def find(x: int, P: list[int]) -> int:
  if P[x] < 0 : return x
  P[x] = find(P[x], P)
  return P[x]

def union(x: int, y: int, P: list[int], C: list[list[int]], ans: int) -> int:
  x = find(x, P)
  y = find(y, P)
  if x == y: return ans
  if P[x] == P[y]: P[x]-=1
  if P[x] > P[y]: x, y = y, x

  ans -= C[x][0] * C[x][1] + C[y][0] * C[y][1]
  C[x][0] += C[y][0]
  C[x][1] += C[y][1]
  ans += C[x][0] * C[x][1]

  P[y] = x
  return ans

def main() -> None:
  # 사람 수, 모임 수
  N, M = map(int, input().split())
  
  # Parent info
  P: list[int] = [ -1 for _ in range(N+1) ]
  # 서로 연락할 수 있는 남녀 쌍
  C: list[list[int]] = [ [ 0, 0 ] for _ in range(N+1) ]
  
  # 성별 정보
  G: list[int] = list(map(int, input().split()))

  # 성별 정보에 따라 mapping
  for i in range(len(G)):
    if G[i] % 2 == 0: C[i+1][0] = 1
    else: C[i+1][1] = 1

    
  # solve & output
  ans: int = 0
  for _ in range(M):
    u, v = map(int, input().split())
    ans = union(u, v, P, C, ans)
    print(ans)


sys.exit(main())

File: 백준/Gold/32203. 연락/연락.java
import java.io.*;
import java.util.*;

public class Main {
  public static int N, M;
  public static long[][] cnt;
  public static int[] p;
  public static long ans;

  public static int swap(int x, int y){ return x; }

  public static int find(int x){
    if(p[x] < 0) return x;
    return p[x] = find(p[x]);
  }

  public static void union(int x, int y){
    x = find(x);
    y = find(y);
    if(x == y) return;
    if(p[x] == p[y]) p[x]--;
    if(p[x] > p[y]) x = swap(y, y = x);

    ans -= cnt[x][0] * cnt[x][1] + cnt[y][0] * cnt[y][1];
    cnt[x][0] += cnt[y][0];
    cnt[x][1] += cnt[y][1];
    ans += cnt[x][0] * cnt[x][1];

    p[y] = x;
  }

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    StringTokenizer st = new StringTokenizer(br.readLine());

    // input 1
    N = Integer.parseInt(st.nextToken()); 
    M = Integer.parseInt(st.nextToken());

    // init
    cnt = new long[N+1][2];
    p = new int[N+1];
    for(int i = 0; i <= N; i++) p[i] = -1;

    // input 2
    st = new StringTokenizer(br.readLine());
    for(int cur = 1; cur <= N; cur++){
      int G = Integer.parseInt(st.nextToken());
      if(G % 2 == 0) cnt[cur][0]++;
      else cnt[cur][1]++;
    }

    // solve
    while(M-- > 0){
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());
      union(u, v);
      sb.append(ans+"\n"); 
    }

    // output
    System.out.println(sb.toString());
  }
}